src/util/animate.ts

```diff
@@ -1,34 +1,124 @@
-//@ts-nocheck
 import { fabric } from '../../HEADER';
 import { runningAnimations } from './animation_registry';
 import { noop } from '../constants';
+import { WithReturnType } from '../typedefs';
+import { defaultEasing, EasingFunction } from './anim_ease';
 
 /**
- *
- * @typedef {Object} AnimationOptions
- * Animation of a value or list of values.
- * @property {Function} [onChange] Callback; invoked on every value change
- * @property {Function} [onComplete] Callback; invoked when value change is completed
- * @property {number | number[]} [startValue=0] Starting value
- * @property {number | number[]} [endValue=100] Ending value
- * @property {number | number[]} [byValue=100] Value to modify the property by
- * @property {Function} [easing] Easing function
- * @property {number} [duration=500] Duration of change (in ms)
- * @property {Function} [abort] Additional function with logic. If returns true, animation aborts.
- * @property {number} [delay] Delay of animation start (in ms)
- *
- * @typedef {() => void} CancelFunction
- *
- * @typedef {Object} AnimationCurrentState
- * @property {number | number[]} currentValue value in range [`startValue`, `endValue`]
- * @property {number} completionRate value in range [0, 1]
- * @property {number} durationRate value in range [0, 1]
- *
- * @typedef {(AnimationOptions & AnimationCurrentState & { cancel: CancelFunction }} AnimationContext
+ * Callback called every frame
+ * @param t current "time"/ms elapsed. multivalue
+ * @param valueRatio ratio of current value to animation max value. [0, 1]
+ * @param timeRatio ratio of current ms to animation duration. [0, 1]
  */
+export type OnChangeCallback = (
+  t: number | number[],
+  valueRatio: number,
+  timeRatio: number
+) => void;
+
+/**
+ * Called to determine if animation should abort
+ * @returns truthy if animation should abort
+ */
+export type AbortCallback = WithReturnType<OnChangeCallback, boolean>;
+
+/**
+ * Function used for canceling an animation
+ */
+export type CancelFunction = VoidFunction;
+
+/**
+ * Animation of a value or list of values
+ */
+export interface AnimationOptions {
+  /**
+   * The object this animation is being performed on
+   */
+  target?: unknown;
```

Comment:
is this referenced anywhere? i can't find it used in the code

Reply:
It is referenced in animation.minix.ts _animate function and the animation unit tests, as well as the animation_registry.ts.

Reply:
@Lazauya is correct, it is used by the animation registry to `cancelByTarget`

Reply:
types should be adapted there as well and liked to here

Reply:
d2075e6

Reply:
> > there are at least 3 behavor change i would like to see reverted, and it seems to me you added a `target` new option that should be proposed on a separate PR.
> 
> The only ACTUAL behavior change is that the currentState of a color animation is now a TColorAlphaSource instead of a color string. See my other comments about the target stuff. I can change it but the problem is its SUPER hard to generify the `animate` util function to work with "normal" animations and color animations. It took me a long time to realize that just changing `animateColor` match `animate` would be more clean in the end. I can make another PR specifically for that I suppose if that would be better.

I think the 3 behavor change are:

- removal of `...restOptions`
- remove of calculateColor where @ShaMan123 pointed out
- on the onChange part assuming that the value is always an array and removing the array check

Arent' those changes?

Adding in pr description `modified the color animate function to be more generic.` doesn't help much in understanding why are you changing things, on top of that those changes should be factored away from a typing pr.